### PR TITLE
Handle Science Museum errors with batches larger than 50 pages

### LINKED ISSUE #864: Science museum script needs to implement `get_should_continue` to stop ingestion
## Description

The ingestion for Science museum doesn't stop when the last page is reached, which causes an error due to response 400 for non-existent page.

## Reproduction

1.  Run an API request for page 50: 

```
curl -X GET -H 'Accept: application/json' -v -i 'https://collection.sciencemuseumgroup.org.uk/search/?has_image=1&image_license=CC&page%5Bsize%5D=100&page%5Bnumber%5D=50&date%5Bfrom%5D=1750&date%5Bto%5D=1825'
``` 

2.  Run a new API request, updating the URL's `page number` parameter to 51:

```
curl -X GET -H 'Accept: application/json' -v -i 'https://collection.sciencemuseumgroup.org.uk/search/?has_image=1&image_license=CC&page%5Bsize%5D=100&page%5Bnumber%5D=51&date%5Bfrom%5D=1750&date%5Bto%5D=1825'
```
3. See error: you will get a 400 response.

## Solution
Each JSON response has a `links` dictionary with `next` property in it. We should stop the ingestion when `next` is `null` by overriding the `get_should_continue` function:

```python
def get_should_continue(self, response_json):
   return response_json.get('links', {}).get('next') is not None
```

## Resolution

- [ ] 🙋 I would be interested in resolving this bug.

### PR DESCRIPTION
**Pull request by stacimc**
## Fixes

Fixes #864

## Background

The Science Museum API appears to be throwing 400 errors when you attempt to access a page number greater than 50 for large datasets, even when there should legitimately be > 50 pages. For example, a recent run received this response for the 50th page of a dataset:

```
# Excerpt from response_json
  "links":{
      "self":"https://collection.sciencemuseumgroup.org.uk/search/date[from]/1875/date[to]/1900/images/image_license?page[size]=100&page[number]=50",
      "first":"https://collection.sciencemuseumgroup.org.uk/search/date[from]/1875/date[to]/1900/images/image_license?page[size]=100",
      "last":"https://collection.sciencemuseumgroup.org.uk/search/date[from]/1875/date[to]/1900/images/image_license?page[size]=100&page[number]=65",
      "prev":"https://collection.sciencemuseumgroup.org.uk/search/date[from]/1875/date[to]/1900/images/image_license?page[size]=100&page[number]=49",
      "next":"https://collection.sciencemuseumgroup.org.uk/search/date[from]/1875/date[to]/1900/images/image_license?page[size]=100&page[number]=51"
   },
   "meta":{
      "total_pages":66,
      ...
```

Note that `total_pages` > 50, and the `links` section _does contain_ a link to page 51. However if you attempt to actually request page 51, you get a 400 error and the DAG eventually errors after a few retry attempts.

We were already splitting the data from the Science Museum up into `year_range` intervals to try to get smaller batches of data, presumably to avoid this error. It looks as though some of the year ranges are now too large.

## Description
This PR:
* Reduces the length of our existing `year_range` intervals so that none contains more than 50 pages of data
* Updates the year_range calculation to be future-proof
* Future-proofs the DAG by detecting when we're about to exceed 50 pages, then halting ingestion for the batch early and alerting in Slack.

In order to get the Slack alert working I had to make a small modification to pass `dag_id` into provider data ingesters, so there are a few changes related to that. I also updated docs and cleaned up a few small things.

## Testing Instructions

`just test`

Run the DAG locally and ensure that you don't see any Slack alerts or errors. Setting an `ingestion_limit` will prevent the buggy scenario from being reached, so if you want to speed up the process you can instead manually edit the starting `page_number` in `get_next_query_params`:

```
    from_, to_ = kwargs["year_range"]
        if not prev_query_params:
            # Reset the page number to 49 for testing
            self.page_number = 49
        else:
            # Increment the page number
            self.page_number += 1
```

To test the Slack alerting when a page number greater than 50 is reached, you can edit the `year_ranges` locally to be a few very long ranges that will definitely have more than 50 pages:
```
year_ranges = [(0, 1900), (1900, 2023),]
```

Then run the DAG again and verify that you can see the `send_alert` message in the logs, but ingestion continues. (Ie you should ingest the first 50 pages worth of data for each of your long ranges).

## Checklist

- [ ] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [ ] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [ ] My commit messages follow [best practices][best_practices].
- [ ] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by AetherUnbound** [comment]
In doing some poking around on the Science Museum side of things, I was able to find this issue: https://github.com/TheScienceMuseum/collectionsonline/issues/735. Looks like they came to a similar conclusion as us: https://github.com/WordPress/openverse-api/issues/857

Pagination is a hard problem it seems!

What's interesting is that their API limits to the first 150 pages still: 

https://github.com/TheScienceMuseum/collectionsonline/blob/6e29580414256db34783993a97624d75d60e9b8d/lib/search.js#L15

Do you think we should reach out for this case and file a bug on their repo, given that the issue we're seeing is on page 50 and not 150? And also given that a `next` link _is still provided_ even though we aren't allowed to traverse beyond page 50.

**Review by AetherUnbound — APPROVED** [review]
This is excellent! I'm not able to test it right now, but the code looks good. love the new tests 🙌🏼 Just a few notes on wording and my previous comments wondering if we should file an upstream issue.

**Comment by AetherUnbound** [comment]
I was able to test, this worked well! One note, your example has: `year_ranges = [(0-1900), (1900-2023),]`, I think it needs to be `year_ranges = [(0, 1900), (1900, 2023),]` (I was getting a `TypeError` when I copy and pasted the example since it was just subtracting the numbers).

File: openverse_catalog/dags/providers/provider_api_scripts/science_museum.py [hunk 1]
```diff
@@ -40,31 +38,58 @@ def __init__(self, *args, **kwargs):
...
 
+        # track page number as instance variable so we can more
+        # easily detect when the page limit is reached
+        self.page_number = 0
+
+    @staticmethod
+    def _get_year_ranges(final_year):
+        """
+        The Science Museum API raises a 400 when attempting to access any
+        page number higher than 50.
```
[the remark this answers is not in the document]
Comment (krysal) [reply openverse_catalog/dags/providers/provider_api_scripts/science_museum.py:49]: Is the Science Museum API open source too? 😮 That would be rad!

**Review by krysal — APPROVED** [review]
LGTM! Lots of nice improvements here.

File: openverse_catalog/dags/providers/provider_api_scripts/science_museum.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (stacimc) [reply openverse_catalog/dags/providers/provider_api_scripts/science_museum.py]: Agreed, done!

File: openverse_catalog/dags/providers/provider_api_scripts/science_museum.py [hunk 2]
```diff
@@ -224,11 +249,29 @@ def _get_license(image_data) -> None | tuple[str, str]:
...
+        # external API errors caused by accessing a page number higher
+        # than 50.
+        if self.page_number == 50:
+            current_url = response_json.get("links", {}).get("self")
+            message = (
+                f"Request {current_url} contains more than 50 pages of"
+                " data and cannot be fully ingested. Reduce the size of"
+                " the year range in order to complete ingestion."
+            )
+            slack.send_alert(message, dag_id=self.dag_id)
```
[the remark this answers is not in the document]
Comment (krysal) [reply openverse_catalog/dags/providers/provider_api_scripts/science_museum.py:269]: Looking again, it's in the logs! The URL is so long that I have to scroll horizontally to see the message. All good 👍

**Pull request merged**